ultralytics/yolo/utils/callbacks/comet.py

```diff
@@ -10,41 +13,319 @@
+    box = ops.xyxy2xywh(box)
+    # adjust xy center to correspond top-left corner
+    box[:2] -= box[2:] / 2
+    box = box.tolist()
+
+    return box
+
+
+def _format_ground_truth_annotations_for_detection(img_idx, image_path, batch, class_name_map=None):
+    indices = batch['batch_idx'] == img_idx
+    bboxes = batch['bboxes'][indices]
+    if len(bboxes) == 0:
+        LOGGER.debug(f'COMET WARNING: Image: {image_path} has no bounding boxes labels')
+        return None
+
+    cls_labels = batch['cls'][indices].squeeze(1).tolist()
+    if class_name_map:
+        cls_labels = [str(class_name_map[label]) for label in cls_labels]
+
+    original_image_shape = batch['ori_shape'][img_idx]
+    resized_image_shape = batch['resized_shape'][img_idx]
+    ratio_pad = batch['ratio_pad'][img_idx]
+
+    data = []
+    for box, label in zip(bboxes, cls_labels):
+        box = _scale_bounding_box_to_original_image_shape(box, resized_image_shape, original_image_shape, ratio_pad)
+        data.append({'boxes': [box], 'label': f'gt_{label}', 'score': COMET_MAX_CONFIDENCE_SCORE})
+
+    return {'name': 'ground_truth', 'data': data}
+
+
+def _format_prediction_annotations_for_detection(image_path, metadata, class_label_map=None):
+    stem = image_path.stem
+    image_id = int(stem) if stem.isnumeric() else stem
+
+    predictions = metadata.get(image_id)
+    if not predictions:
+        LOGGER.debug(f'COMET WARNING: Image: {image_path} has no bounding boxes predictions')
+        return None
+
+    data = []
+    for prediction in predictions:
+        boxes = prediction['bbox']
+        score = prediction['score'] * COMET_MAX_CONFIDENCE_SCORE
+        cls_label = prediction['category_id']
+        if class_label_map:
+            cls_label = str(class_label_map[cls_label])
+
+        data.append({'boxes': [boxes], 'label': cls_label, 'score': score})
+
+    return {'name': 'prediction', 'data': data}
+
+
+def _fetch_annotations(img_idx, image_path, batch, prediction_metadata_map, class_label_map):
+    ground_truth_annotations = _format_ground_truth_annotations_for_detection(img_idx, image_path, batch,
+                                                                              class_label_map)
+    prediction_annotations = _format_prediction_annotations_for_detection(image_path, prediction_metadata_map,
+                                                                          class_label_map)
+
+    annotations = [
+        annotation for annotation in [ground_truth_annotations, prediction_annotations] if annotation is not None]
+    return [annotations] if annotations else None
+
+
+def _create_prediction_metadata_map(model_predictions):
+    pred_metadata_map = {}
+    for prediction in model_predictions:
+        pred_metadata_map.setdefault(prediction['image_id'], [])
+        pred_metadata_map[prediction['image_id']].append(prediction)
+
+    return pred_metadata_map
+
+
+def _log_confusion_matrix(experiment, trainer, curr_step, curr_epoch):
```

Comment:
@DN6 unless I'm missing something here, this Confusion matrix will be incorrect as you are sorting class names alphabetically when the validator confusion matrix has been created with a class index sorting. This means you are detaching the class names from their corresponding index.